File: apps/server/src/db/drivers/mysql/introspect.ts
import type mysql from 'mysql2/promise'
import type {
  DatabaseInfo,
  SchemaTree,
  ColumnInfo,
  IndexInfo,
  ConstraintInfo,
  CompletionColumn,
  CompletionFunction,
  CompletionForeignKey,
  TableDataOptions,
} from '$src/db/types'
import type { EditableSource } from '$src/db/driver'
import type { ColumnMeta, QueryResult } from '$src/db/shared/marshal'
import { executeSql } from '$src/db/drivers/mysql/exec'
import { quoteIdent } from '$src/db/drivers/mysql/dml'

const SYSTEM_SCHEMAS = ["'information_schema'", "'performance_schema'", "'sys'"]

function formatBytes(bytes: number): string {
  if (!bytes || bytes === 0) return '0 B'
  const k = 1024
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB']
  const i = Math.floor(Math.log(bytes) / Math.log(k))
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(1))} ${sizes[i]}`
}

export async function listDatabases(pool: mysql.Pool): Promise<DatabaseInfo[]> {
  const sql = `
    SELECT
      s.SCHEMA_NAME AS name,
      'root' AS owner,
      COALESCE(SUM(t.DATA_LENGTH + t.INDEX_LENGTH), 0) AS sizeBytes
    FROM information_schema.SCHEMATA s
    LEFT JOIN information_schema.TABLES t ON s.SCHEMA_NAME = t.TABLE_SCHEMA
    GROUP BY s.SCHEMA_NAME
    ORDER BY s.SCHEMA_NAME
  `
  const [rows] = await pool.query({ sql, rowsAsArray: false })
  return (rows as Array<{ name: string; owner: string; sizeBytes: number | string }>).map((r) => ({
    name: r.name,
    owner: r.owner,
    size: formatBytes(Number(r.sizeBytes) || 0),
  }))
}

export async function listSchemas(pool: mysql.Pool): Promise<string[]> {
  const sql = `
    SELECT SCHEMA_NAME AS name
    FROM information_schema.SCHEMATA
    WHERE SCHEMA_NAME NOT IN (${SYSTEM_SCHEMAS.join(', ')})
    ORDER BY SCHEMA_NAME
  `
  const [rows] = await pool.query({ sql, rowsAsArray: false })
  return (rows as Array<{ name: string }>).map((r) => r.name)
}

export async function getTree(pool: mysql.Pool, search?: string): Promise<SchemaTree[]> {
  let sql = `
    SELECT
      TABLE_SCHEMA AS \`schema\`,
      TABLE_NAME AS \`name\`,
      CASE WHEN TABLE_TYPE = 'VIEW' THEN 'view' ELSE 'table' END AS \`type\`
    FROM information_schema.TABLES
    WHERE TABLE_SCHEMA NOT IN (${SYSTEM_SCHEMAS.join(', ')})
  `
  const params: unknown[] = []
  if (search && search.trim()) {
    sql += ` AND (TABLE_NAME LIKE ? OR TABLE_SCHEMA LIKE ?)`
    const pattern = `%${search.trim()}%`
    params.push(pattern, pattern)
  }
  sql += ` ORDER BY TABLE_SCHEMA, TABLE_NAME`

  const [rows] = await pool.query({ sql, rowsAsArray: false, values: params })
  const grouped: Record<string, { name: string; type: 'table' | 'view' | 'matview' }[]> = {}

  for (const r of rows as Array<{ schema: string; name: string; type: 'table' | 'view' }>) {
    if (!grouped[r.schema]) grouped[r.schema] = []
    grouped[r.schema].push({ name: r.name, type: r.type })
  }

  return Object.entries(grouped).map(([schema, relations]) => ({ schema, relations }))
}

function parseEnumValues(columnType: string): string[] | null {
  const match = columnType.match(/^enum\((.*)\)$/i)
  if (!match) return null
  return match[1].split(',').map((v) =>
    v
      .trim()
      .replace(/^'(.*)'$/, '$1')
      .replace(/''/g, "'"),
  )
}

export async function getColumns(
  pool: mysql.Pool,
  schema: string | null,
  table: string,
): Promise<ColumnInfo[]> {
  const sql = `
    SELECT
      COLUMN_NAME AS name,
      DATA_TYPE AS type,
      COLUMN_TYPE AS columnType,
      IS_NULLABLE AS isNullable,
      COLUMN_DEFAULT AS \`default\`,
      ORDINAL_POSITION AS ordinal,
      COLUMN_KEY AS columnKey,
      EXTRA AS extra
    FROM information_schema.COLUMNS
    WHERE TABLE_SCHEMA = COALESCE(?, DATABASE()) AND TABLE_NAME = ?
    ORDER BY ORDINAL_POSITION
  `
  const [rows] = await pool.query({ sql, rowsAsArray: false, values: [schema, table] })
  return (
    rows as Array<{
      name: string
      type: string
      columnType: string
      isNullable: string
      default: string | null
      ordinal: number
      columnKey: string
      extra: string
    }>
  ).map((r) => ({
    name: r.name,
    type: r.columnType || r.type,
    notNull: r.isNullable === 'NO',
    default: r.default,
    ordinal: Number(r.ordinal),
    isPrimaryKey: r.columnKey === 'PRI',
    enumValues: parseEnumValues(r.columnType),
  }))
}

export async function getIndexes(
  pool: mysql.Pool,
  schema: string | null,
  table: string,
): Promise<IndexInfo[]> {
  const sql = `
    SELECT
      INDEX_NAME AS name,
      NON_UNIQUE AS nonUnique,
      INDEX_TYPE AS indexType,
      GROUP_CONCAT(COLUMN_NAME ORDER BY SEQ_IN_INDEX) AS columns
    FROM information_schema.STATISTICS
    WHERE TABLE_SCHEMA = COALESCE(?, DATABASE()) AND TABLE_NAME = ?
    GROUP BY INDEX_NAME, NON_UNIQUE, INDEX_TYPE
  `
  const [rows] = await pool.query({ sql, rowsAsArray: false, values: [schema, table] })
  return (
    rows as Array<{
      name: string
      nonUnique: number
      indexType: string
      columns: string
    }>
  ).map((r) => {
    const isPrimary = r.name === 'PRIMARY'
    const isUnique = Number(r.nonUnique) === 0
    const cols = r.columns ? r.columns.split(',').map(quoteIdent).join(', ') : ''
    const definition = isPrimary
      ? `PRIMARY KEY (${cols})`
      : `${isUnique ? 'UNIQUE ' : ''}INDEX ${quoteIdent(r.name)} (${cols}) USING ${r.indexType}`
    return {
      name: r.name,
      definition,
      isUnique,
      isPrimary,
    }
  })
}

export async function getConstraints(
  pool: mysql.Pool,
  schema: string | null,
  table: string,
): Promise<ConstraintInfo[]> {
  const sql = `
    SELECT
      tc.CONSTRAINT_NAME AS name,
      tc.CONSTRAINT_TYPE AS type,
      kcu.COLUMN_NAME AS columnName,
      kcu.REFERENCED_TABLE_SCHEMA AS refSchema,
      kcu.REFERENCED_TABLE_NAME AS refTable,
      kcu.REFERENCED_COLUMN_NAME AS refColumn
    FROM information_schema.TABLE_CONSTRAINTS tc
    LEFT JOIN information_schema.KEY_COLUMN_USAGE kcu
      ON tc.CONSTRAINT_SCHEMA = kcu.CONSTRAINT_SCHEMA
      AND tc.TABLE_NAME = kcu.TABLE_NAME
      AND tc.CONSTRAINT_NAME = kcu.CONSTRAINT_NAME
    WHERE tc.TABLE_SCHEMA = COALESCE(?, DATABASE()) AND tc.TABLE_NAME = ?
    ORDER BY tc.CONSTRAINT_NAME, kcu.ORDINAL_POSITION
  `
  const [rows] = await pool.query({ sql, rowsAsArray: false, values: [schema, table] })
  return (
    rows as Array<{
      name: string
      type: string
      columnName: string | null
      refSchema: string | null
      refTable: string | null
      refColumn: string | null
    }>
  ).map((r) => {
    let definition = `${r.type} (${r.columnName ? quoteIdent(r.columnName) : ''})`
    if (r.type === 'FOREIGN KEY' && r.refTable && r.refColumn) {
      definition += ` REFERENCES ${r.refSchema ? `${quoteIdent(r.refSchema)}.` : ''}${quoteIdent(r.refTable)} (${quoteIdent(r.refColumn)})`
    }
    return {
      name: r.name,
      type: r.type,
      definition,
    }
  })
}

export async function getPrimaryKey(
  pool: mysql.Pool,
  schema: string | null,
  table: string,
): Promise<string[]> {
  const sql = `
    SELECT COLUMN_NAME AS name
    FROM information_schema.KEY_COLUMN_USAGE
    WHERE CONSTRAINT_NAME = 'PRIMARY'
      AND TABLE_SCHEMA = COALESCE(?, DATABASE())
      AND TABLE_NAME = ?
    ORDER BY ORDINAL_POSITION
  `
  const [rows] = await pool.query({ sql, rowsAsArray: false, values: [schema, table] })
  return (rows as Array<{ name: string }>).map((r) => r.name)
}

export async function getAllColumns(pool: mysql.Pool): Promise<CompletionColumn[]> {
  const sql = `
    SELECT
      TABLE_SCHEMA AS \`schema\`,
      TABLE_NAME AS \`table\`,
      COLUMN_NAME AS \`name\`,
      DATA_TYPE AS \`type\`
    FROM information_schema.COLUMNS
    WHERE TABLE_SCHEMA NOT IN (${SYSTEM_SCHEMAS.join(', ')})
  `
  const [rows] = await pool.query({ sql, rowsAsArray: false })
  return rows as CompletionColumn[]
}

export async function getFunctions(pool: mysql.Pool): Promise<CompletionFunction[]> {
  const sql = `
    SELECT
      ROUTINE_SCHEMA AS \`schema\`,
      ROUTINE_NAME AS \`name\`,
      COALESCE(DTD_IDENTIFIER, DATA_TYPE, '') AS \`returns\`,
      '' AS \`args\`,
      LOWER(ROUTINE_TYPE) AS \`kind\`
    FROM information_schema.ROUTINES
    WHERE ROUTINE_SCHEMA NOT IN (${SYSTEM_SCHEMAS.join(', ')})
  `
  const [rows] = await pool.query({ sql, rowsAsArray: false })
  return (
    rows as Array<{ schema: string; name: string; returns: string; args: string; kind: string }>
  ).map((r) => ({
    schema: r.schema,
    name: r.name,
    returns: r.returns,
    args: r.args,
    kind: (r.kind === 'procedure' ? 'procedure' : 'function') as 'function' | 'procedure',
  }))
}

export async function getForeignKeys(pool: mysql.Pool): Promise<CompletionForeignKey[]> {
  const sql = `
    SELECT
      CONSTRAINT_NAME AS name,
      TABLE_SCHEMA AS fromSchema,
      TABLE_NAME AS fromTable,
      COLUMN_NAME AS fromColumn,
      REFERENCED_TABLE_SCHEMA AS refSchema,
      REFERENCED_TABLE_NAME AS refTable,
      REFERENCED_COLUMN_NAME AS refColumn
    FROM information_schema.KEY_COLUMN_USAGE
    WHERE REFERENCED_TABLE_NAME IS NOT NULL
      AND TABLE_SCHEMA NOT IN (${SYSTEM_SCHEMAS.join(', ')})
    ORDER BY CONSTRAINT_NAME, ORDINAL_POSITION
  `
  const [rows] = await pool.query({ sql, rowsAsArray: false })
  const map = new Map<string, CompletionForeignKey>()

  for (const r of rows as Array<{
    name: string
    fromSchema: string
    fromTable: string
    fromColumn: string
    refSchema: string
    refTable: string
    refColumn: string
  }>) {
    const key = `${r.fromSchema}.${r.fromTable}.${r.name}`
    const existing = map.get(key)
    if (existing) {
      existing.columns.push(r.fromColumn)
      existing.refColumns.push(r.refColumn)
    } else {
      map.set(key, {
        schema: r.fromSchema,
        table: r.fromTable,
        columns: [r.fromColumn],
        refSchema: r.refSchema,
        refTable: r.refTable,
        refColumns: [r.refColumn],
      })
    }
  }

  return Array.from(map.values())
}

export async function getTableData(
  pool: mysql.Pool,
  schema: string | null,
  table: string,
  opts: TableDataOptions,
): Promise<{ result: QueryResult; estimatedRows: number }> {
  const target = schema ? `${quoteIdent(schema)}.${quoteIdent(table)}` : quoteIdent(table)
  let sql = `SELECT * FROM ${target}`
  const params: unknown[] = []

  if (opts.filterColumn && opts.filterValue !== undefined) {
    sql += ` WHERE ${quoteIdent(opts.filterColumn)} = ?`
    params.push(opts.filterValue)
  }

  if (opts.orderBy) {
    sql += ` ORDER BY ${quoteIdent(opts.orderBy)} ${opts.orderDir === 'DESC' ? 'DESC' : 'ASC'}`
  }

  sql += ` LIMIT ? OFFSET ?`
  params.push(opts.limit, opts.offset)

  const result = await executeSql(pool, sql, params)

  // Get estimated rows from information_schema
  const estSql = `
    SELECT TABLE_ROWS AS count
    FROM information_schema.TABLES
    WHERE TABLE_SCHEMA = COALESCE(?, DATABASE()) AND TABLE_NAME = ?
  `
  const [estRows] = await pool.query({ sql: estSql, rowsAsArray: false, values: [schema, table] })
  const estimated = Number((estRows as Array<{ count?: number }>)?.[0]?.count) || result.rowCount

  return { result, estimatedRows: estimated }
}

export async function resolveSource(
  pool: mysql.Pool,
  columns: ColumnMeta[],
): Promise<EditableSource | null> {
  if (columns.length === 0) return null
  // In MySQL, check if columns belong to single table
  // If result has columns without base table info, fallback
  const firstCol = columns[0]
  if (!firstCol) return null

  // If executing a direct SELECT * FROM table, extract from completion/tree
  return null
}

File: crates/geto-core/src/db/drivers/oracle/introspect.rs
use indexmap::IndexMap;

use crate::db::types::{
    ColumnInfo, CompletionColumn, CompletionForeignKey, CompletionFunction,
    CompletionResponse, CompletionTable, ConstraintInfo, DatabaseInfo, IndexInfo,
    RelationEntry, RelationType, SchemaTree, TableDetailResponse,
};
use crate::error::AppError;

fn get_str(row: &oracle_rs::Row, cols: &[oracle_rs::ColumnInfo], col_name: &str) -> Option<String> {
    for (i, c) in cols.iter().enumerate() {
        if c.name.eq_ignore_ascii_case(col_name) {
            if let Some(v) = row.get(i) {
                return match v {
                    oracle_rs::Value::String(s) => Some(s.clone()),
                    oracle_rs::Value::Null => None,
                    other => other.as_str().map(|s| s.to_string()),
                };
            }
        }
    }
    None
}

fn get_i64(row: &oracle_rs::Row, cols: &[oracle_rs::ColumnInfo], col_name: &str) -> Option<i64> {
    for (i, c) in cols.iter().enumerate() {
        if c.name.eq_ignore_ascii_case(col_name) {
            if let Some(v) = row.get(i) {
                return match v {
                    oracle_rs::Value::Integer(n) => Some(*n),
                    oracle_rs::Value::Number(num) => num.to_i64().ok(),
                    _ => v.as_i64(),
                };
            }
        }
    }
    None
}

pub async fn list_databases(
    conn: &oracle_rs::Connection,
    default_name: &str,
) -> Result<Vec<DatabaseInfo>, AppError> {
    let sql = "SELECT sys_context('USERENV', 'DB_NAME') AS name FROM dual";
    let qr = conn
        .query(sql, &[])
        .await
        .map_err(|e| AppError::Database(format!("Failed to list Oracle databases: {}", e)))?;

    let name = qr
        .rows
        .first()
        .and_then(|r| get_str(r, &qr.columns, "name"))
        .unwrap_or_else(|| {
            if default_name.is_empty() {
                "ORCL".to_string()
            } else {
                default_name.to_string()
            }
        });

    Ok(vec![DatabaseInfo {
        name,
        owner: "SYS".to_string(),
        size: "Unknown".to_string(),
    }])
}

pub async fn list_schemas(conn: &oracle_rs::Connection) -> Result<Vec<String>, AppError> {
    let sql = "SELECT username AS name FROM all_users ORDER BY username";
    let qr = conn
        .query(sql, &[])
        .await
        .map_err(|e| AppError::Database(format!("Failed to list Oracle schemas: {}", e)))?;

    let mut out = Vec::new();
    for row in &qr.rows {
        if let Some(s) = get_str(row, &qr.columns, "name") {
            out.push(s);
        }
    }
    Ok(out)
}

pub async fn get_tree(
    conn: &oracle_rs::Connection,
    search: Option<&str>,
) -> Result<Vec<SchemaTree>, AppError> {
    let base_sql = r#"
        SELECT owner, table_name AS name, 'table' AS kind
        FROM all_tables
        WHERE owner NOT IN ('SYS', 'SYSTEM', 'OUTLN', 'XDB', 'CTXSYS', 'MDSYS', 'ORDSYS', 'ORDDATA', 'WMSYS', 'LBACSYS', 'DVSYS', 'GSMADMIN_INTERNAL', 'AUDSYS', 'APPQOSSYS', 'DBSNMP')
        UNION ALL
        SELECT owner, view_name AS name, 'view' AS kind
        FROM all_views
        WHERE owner NOT IN ('SYS', 'SYSTEM', 'OUTLN', 'XDB', 'CTXSYS', 'MDSYS', 'ORDSYS', 'ORDDATA', 'WMSYS', 'LBACSYS', 'DVSYS', 'GSMADMIN_INTERNAL', 'AUDSYS', 'APPQOSSYS', 'DBSNMP')
        ORDER BY owner, name
    "#;

    let qr = conn
        .query(base_sql, &[])
        .await
        .map_err(|e| AppError::Database(format!("Failed to introspect Oracle schema tree: {}", e)))?;

    let search_lc = search.map(|s| s.trim().to_lowercase()).filter(|s| !s.is_empty());
    let mut map: IndexMap<String, Vec<RelationEntry>> = IndexMap::new();

    for row in &qr.rows {
        let schema = get_str(row, &qr.columns, "owner").unwrap_or_default();
        let name = get_str(row, &qr.columns, "name").unwrap_or_default();
        let kind = get_str(row, &qr.columns, "kind").unwrap_or_default();

        if let Some(ref pat) = search_lc {
            if !name.to_lowercase().contains(pat) {
                continue;
            }
        }

        let rel_type = match kind.as_str() {
            "view" => RelationType::View,
            _ => RelationType::Table,
        };

        map.entry(schema).or_default().push(RelationEntry {
            name,
            r#type: rel_type,
        });
    }

    let out = map
        .into_iter()
        .map(|(schema, relations)| SchemaTree { schema, relations })
        .collect();

    Ok(out)
}

pub async fn get_columns(
    conn: &oracle_rs::Connection,
    schema: Option<&str>,
    table: &str,
) -> Result<Vec<ColumnInfo>, AppError> {
    let sql = r#"
        SELECT
            c.column_name AS name,
            c.data_type AS type,
            c.nullable,
            c.data_default AS def_val,
            c.column_id AS ordinal,
            CASE WHEN pk.column_name IS NOT NULL THEN 1 ELSE 0 END AS is_pk
        FROM all_tab_columns c
        LEFT JOIN (
            SELECT cc.owner, cc.table_name, cc.column_name
            FROM all_constraints con
            JOIN all_cons_columns cc ON con.constraint_name = cc.constraint_name AND con.owner = cc.owner
            WHERE con.constraint_type = 'P'
        ) pk ON c.owner = pk.owner AND c.table_name = pk.table_name AND c.column_name = pk.column_name
        WHERE UPPER(c.table_name) = UPPER(:1)
          AND (UPPER(c.owner) = UPPER(:2) OR :2 IS NULL)
        ORDER BY c.column_id
    "#;

    let tbl_val = oracle_rs::Value::String(table.to_string());
    let schema_val = schema
        .filter(|s| !s.is_empty())
        .map(|s| oracle_rs::Value::String(s.to_string()))
        .unwrap_or(oracle_rs::Value::Null);

    let qr = conn
        .query(sql, &[tbl_val, schema_val])
        .await
        .map_err(|e| AppError::Database(format!("Failed to get Oracle columns: {}", e)))?;

    let mut out = Vec::with_capacity(qr.rows.len());
    for row in &qr.rows {
        let name = get_str(row, &qr.columns, "name").unwrap_or_default();
        let data_type = get_str(row, &qr.columns, "type").unwrap_or_else(|| "VARCHAR2".to_string());
        let nullable_str = get_str(row, &qr.columns, "nullable").unwrap_or_else(|| "Y".to_string());
        let default_val = get_str(row, &qr.columns, "def_val");
        let ordinal = get_i64(row, &qr.columns, "ordinal").unwrap_or(0) as i32;
        let is_pk = get_i64(row, &qr.columns, "is_pk").unwrap_or(0) == 1;

        out.push(ColumnInfo {
            name,
            r#type: data_type,
            not_null: nullable_str == "N",
            default: default_val,
            ordinal,
            is_primary_key: is_pk,
            enum_values: None,
        });
    }

    Ok(out)
}

pub async fn get_indexes(
    conn: &oracle_rs::Connection,
    schema: Option<&str>,
    table: &str,
) -> Result<Vec<IndexInfo>, AppError> {
    let sql = r#"
        SELECT
            i.index_name AS name,
            i.uniqueness,
            c.column_name,
            c.column_position,
            c.descend
        FROM all_indexes i
        JOIN all_ind_columns c ON i.index_name = c.index_name AND i.owner = c.index_owner
        WHERE UPPER(i.table_name) = UPPER(:1)
          AND (UPPER(i.owner) = UPPER(:2) OR :2 IS NULL)
        ORDER BY i.index_name, c.column_position
    "#;

    let tbl_val = oracle_rs::Value::String(table.to_string());
    let schema_val = schema
        .filter(|s| !s.is_empty())
        .map(|s| oracle_rs::Value::String(s.to_string()))
        .unwrap_or(oracle_rs::Value::Null);

    let qr = conn
        .query(sql, &[tbl_val, schema_val])
        .await
        .map_err(|e| AppError::Database(format!("Failed to get Oracle indexes: {}", e)))?;

    let mut map: IndexMap<String, (bool, Vec<String>)> = IndexMap::new();

    for row in &qr.rows {
        let name = get_str(row, &qr.columns, "name").unwrap_or_default();
        let uniqueness = get_str(row, &qr.columns, "uniqueness").unwrap_or_default();
        let col = get_str(row, &qr.columns, "column_name").unwrap_or_default();
        let is_unique = uniqueness == "UNIQUE";

        let entry = map.entry(name).or_insert_with(|| (is_unique, Vec::new()));
        entry.1.push(col);
    }

    let out = map
        .into_iter()
        .map(|(name, (is_unique, columns))| IndexInfo {
            name,
            definition: format!("INDEX ({})", columns.join(", ")),
            is_unique,
            is_primary: false,
        })
        .collect();

    Ok(out)
}

pub async fn get_constraints(
    conn: &oracle_rs::Connection,
    schema: Option<&str>,
    table: &str,
) -> Result<Vec<ConstraintInfo>, AppError> {
    let sql = r#"
        SELECT
            con.constraint_name AS name,
            con.constraint_type AS type,
            cc.column_name,
            r_con.table_name AS ref_table,
            r_cc.column_name AS ref_column
        FROM all_constraints con
        JOIN all_cons_columns cc ON con.constraint_name = cc.constraint_name AND con.owner = cc.owner
        LEFT JOIN all_constraints r_con ON con.r_constraint_name = r_con.constraint_name AND con.r_owner = r_con.owner
        LEFT JOIN all_cons_columns r_cc ON r_con.constraint_name = r_cc.constraint_name AND r_con.owner = r_cc.owner
        WHERE UPPER(con.table_name) = UPPER(:1)
          AND (UPPER(con.owner) = UPPER(:2) OR :2 IS NULL)
        ORDER BY con.constraint_name, cc.position
    "#;

    let tbl_val = oracle_rs::Value::String(table.to_string());
    let schema_val = schema
        .filter(|s| !s.is_empty())
        .map(|s| oracle_rs::Value::String(s.to_string()))
        .unwrap_or(oracle_rs::Value::Null);

    let qr = conn
        .query(sql, &[tbl_val, schema_val])
        .await
        .map_err(|e| AppError::Database(format!("Failed to get Oracle constraints: {}", e)))?;

    let mut map: IndexMap<String, (String, Vec<String>, Option<String>, Vec<String>)> = IndexMap::new();

    for row in &qr.rows {
        let name = get_str(row, &qr.columns, "name").unwrap_or_default();
        let ctype = get_str(row, &qr.columns, "type").unwrap_or_default();
        let col = get_str(row, &qr.columns, "column_name").unwrap_or_default();
        let ref_tbl = get_str(row, &qr.columns, "ref_table");
        let ref_col = get_str(row, &qr.columns, "ref_column");

        let entry = map
            .entry(name)
            .or_insert_with(|| (ctype, Vec::new(), ref_tbl, Vec::new()));
        entry.1.push(col);
        if let Some(rc) = ref_col {
            entry.3.push(rc);
        }
    }

    let out = map
        .into_iter()
        .map(|(name, (ctype, columns, ref_tbl, ref_cols))| {
            let mapped_type = match ctype.as_str() {
                "P" => "PRIMARY KEY",
                "U" => "UNIQUE",
                "R" => "FOREIGN KEY",
                "C" => "CHECK",
                _ => "CHECK",
            };
            let definition = if let Some(ref_t) = ref_tbl {
                format!(
                    "{} ({}) REFERENCES {} ({})",
                    mapped_type,
                    columns.join(", "),
                    ref_t,
                    ref_cols.join(", ")
                )
            } else {
                format!("{} ({})", mapped_type, columns.join(", "))
            };
            ConstraintInfo {
                name,
                r#type: mapped_type.to_string(),
                definition,
            }
        })
        .collect();

    Ok(out)
}

pub async fn get_table_foreign_keys(
    conn: &oracle_rs::Connection,
    schema: Option<&str>,
    table: &str,
) -> Result<Vec<CompletionForeignKey>, AppError> {
    let sql = r#"
        SELECT
            con.owner AS schema,
            con.table_name AS table_name,
            cc.column_name,
            r_con.owner AS ref_schema,
            r_con.table_name AS ref_table,
            r_cc.column_name AS ref_column
        FROM all_constraints con
        JOIN all_cons_columns cc ON con.constraint_name = cc.constraint_name AND con.owner = cc.owner
        JOIN all_constraints r_con ON con.r_constraint_name = r_con.constraint_name AND con.r_owner = r_con.owner
        JOIN all_cons_columns r_cc ON r_con.constraint_name = r_cc.constraint_name AND r_con.owner = r_cc.owner
        WHERE con.constraint_type = 'R'
          AND UPPER(con.table_name) = UPPER(:1)
          AND (UPPER(con.owner) = UPPER(:2) OR :2 IS NULL)
        ORDER BY con.constraint_name, cc.position
    "#;

    let tbl_val = oracle_rs::Value::String(table.to_string());
    let schema_val = schema
        .filter(|s| !s.is_empty())
        .map(|s| oracle_rs::Value::String(s.to_string()))
        .unwrap_or(oracle_rs::Value::Null);

    let qr = conn
        .query(sql, &[tbl_val, schema_val])
        .await
        .map_err(|e| AppError::Database(format!("Failed to get Oracle foreign keys: {}", e)))?;

    let mut out = Vec::with_capacity(qr.rows.len());
    for row in &qr.rows {
        let sch = get_str(row, &qr.columns, "schema").unwrap_or_default();
        let tbl = get_str(row, &qr.columns, "table_name").unwrap_or_else(|| table.to_string());
        let column = get_str(row, &qr.columns, "column_name").unwrap_or_default();
        let ref_sch = get_str(row, &qr.columns, "ref_schema").unwrap_or_default();
        let ref_table = get_str(row, &qr.columns, "ref_table").unwrap_or_default();
        let ref_column = get_str(row, &qr.columns, "ref_column").unwrap_or_default();

        out.push(CompletionForeignKey {
            schema: sch,
            table: tbl,
            columns: vec![column],
            ref_schema: ref_sch,
            ref_table,
            ref_columns: vec![ref_column],
        });
    }

    Ok(out)
}

pub async fn get_table_detail(
    conn: &oracle_rs::Connection,
    schema: Option<&str>,
    table: &str,
) -> Result<TableDetailResponse, AppError> {
    let columns = get_columns(conn, schema, table).await?;
    let indexes = get_indexes(conn, schema, table).await?;
    let constraints = get_constraints(conn, schema, table).await?;
    let foreign_keys = get_table_foreign_keys(conn, schema, table).await?;

    let primary_key = columns
        .iter()
        .filter(|c| c.is_primary_key)
        .map(|c| c.name.clone())
        .collect();

    Ok(TableDetailResponse {
        columns,
        indexes,
        constraints,
        primary_key,
        foreign_keys,
    })
}

pub async fn get_all_columns(conn: &oracle_rs::Connection) -> Result<Vec<CompletionColumn>, AppError> {
    let sql = r#"
        SELECT
            c.owner AS schema,
            c.table_name AS table_name,
            c.column_name AS name,
            c.data_type AS type
        FROM all_tab_columns c
        WHERE c.owner NOT IN ('SYS', 'SYSTEM', 'OUTLN', 'XDB', 'CTXSYS', 'MDSYS')
        ORDER BY c.owner, c.table_name, c.column_id
    "#;

    let qr = conn
        .query(sql, &[])
        .await
        .map_err(|e| AppError::Database(format!("Failed to get all columns: {}", e)))?;

    let mut out = Vec::with_capacity(qr.rows.len());
    for row in &qr.rows {
        let schema = get_str(row, &qr.columns, "schema").unwrap_or_default();
        let table = get_str(row, &qr.columns, "table_name").unwrap_or_default();
        let name = get_str(row, &qr.columns, "name").unwrap_or_default();
        let data_type = get_str(row, &qr.columns, "type").unwrap_or_default();

        out.push(CompletionColumn {
            schema,
            table,
            name,
            r#type: data_type,
        });
    }

    Ok(out)
}

pub async fn get_functions(conn: &oracle_rs::Connection) -> Result<Vec<CompletionFunction>, AppError> {
    let sql = r#"
        SELECT
            object_name AS name,
            owner AS schema,
            object_type AS type
        FROM all_objects
        WHERE object_type IN ('FUNCTION', 'PROCEDURE', 'PACKAGE')
          AND owner NOT IN ('SYS', 'SYSTEM', 'OUTLN', 'XDB', 'CTXSYS', 'MDSYS')
        ORDER BY owner, object_name
    "#;

    let qr = conn
        .query(sql, &[])
        .await
        .map_err(|e| AppError::Database(format!("Failed to get Oracle functions: {}", e)))?;

    let mut out = Vec::new();
    for row in &qr.rows {
        let name = get_str(row, &qr.columns, "name").unwrap_or_default();
        let schema = get_str(row, &qr.columns, "schema").unwrap_or_default();
        let otype = get_str(row, &qr.columns, "type").unwrap_or_default();

        out.push(CompletionFunction {
            schema,
            name,
            args: String::new(),
            returns: otype,
            kind: "function".to_string(),
        });
    }

    // Built-in functions
    let builtins = [
        ("NVL", "expr1, expr2"),
        ("NVL2", "expr1, expr2, expr3"),
        ("DECODE", "expr, search, result, default"),
        ("COALESCE", "expr1, expr2, ..."),
        ("TO_CHAR", "val, format"),
        ("TO_DATE", "str, format"),
        ("TO_NUMBER", "str, format"),
        ("SYSDATE", ""),
        ("SYSTIMESTAMP", ""),
        ("TRUNC", "date, format"),
        ("SUBSTR", "str, pos, len"),
        ("INSTR", "str, sub, pos, nth"),
        ("LENGTH", "str"),
        ("UPPER", "str"),
        ("LOWER", "str"),
        ("INITCAP", "str"),
        ("LPAD", "str, len, pad"),
        ("RPAD", "str, len, pad"),
        ("LTRIM", "str, set"),
        ("RTRIM", "str, set"),
        ("TRIM", "str"),
        ("REPLACE", "str, search, repl"),
        ("REGEXP_SUBSTR", "str, pattern"),
        ("REGEXP_REPLACE", "str, pattern, repl"),
        ("REGEXP_LIKE", "str, pattern"),
        ("LISTAGG", "measure_expr, delimiter"),
        ("COUNT", "expr"),
        ("SUM", "expr"),
        ("AVG", "expr"),
        ("MIN", "expr"),
        ("MAX", "expr"),
        ("DENSE_RANK", ""),
        ("ROW_NUMBER", ""),
    ];

    for (name, args) in builtins {
        out.push(CompletionFunction {
            schema: "SYS".to_string(),
            name: name.to_string(),
            args: args.to_string(),
            returns: String::new(),
            kind: "function".to_string(),
        });
    }

    Ok(out)
}

pub async fn get_foreign_keys(conn: &oracle_rs::Connection) -> Result<Vec<CompletionForeignKey>, AppError> {
    let sql = r#"
        SELECT
            con.owner AS schema,
            con.table_name,
            cc.column_name,
            r_con.owner AS ref_schema,
            r_con.table_name AS ref_table,
            r_cc.column_name AS ref_column
        FROM all_constraints con
        JOIN all_cons_columns cc ON con.constraint_name = cc.constraint_name AND con.owner = cc.owner
        JOIN all_constraints r_con ON con.r_constraint_name = r_con.constraint_name AND con.r_owner = r_con.owner
        JOIN all_cons_columns r_cc ON r_con.constraint_name = r_cc.constraint_name AND r_con.owner = r_cc.owner
        WHERE con.constraint_type = 'R'
          AND con.owner NOT IN ('SYS', 'SYSTEM', 'OUTLN', 'XDB', 'CTXSYS', 'MDSYS')
        ORDER BY con.constraint_name, cc.position
    "#;

    let qr = conn
        .query(sql, &[])
        .await
        .map_err(|e| AppError::Database(format!("Failed to get completion foreign keys: {}", e)))?;

    let mut out = Vec::with_capacity(qr.rows.len());
    for row in &qr.rows {
        let sch = get_str(row, &qr.columns, "schema").unwrap_or_default();
        let table = get_str(row, &qr.columns, "table_name").unwrap_or_default();
        let column = get_str(row, &qr.columns, "column_name").unwrap_or_default();
        let ref_sch = get_str(row, &qr.columns, "ref_schema").unwrap_or_default();
        let ref_table = get_str(row, &qr.columns, "ref_table").unwrap_or_default();
        let ref_column = get_str(row, &qr.columns, "ref_column").unwrap_or_default();

        out.push(CompletionForeignKey {
            schema: sch,
            table,
            columns: vec![column],
            ref_schema: ref_sch,
            ref_table,
            ref_columns: vec![ref_column],
        });
    }

    Ok(out)
}

pub async fn get_completion(conn: &oracle_rs::Connection) -> Result<CompletionResponse, AppError> {
    let tree = get_tree(conn, None).await.unwrap_or_default();

    let mut tables = Vec::new();
    for st in &tree {
        for rel in &st.relations {
            tables.push(CompletionTable {
                schema: st.schema.clone(),
                name: rel.name.clone(),
                r#type: rel.r#type,
            });
        }
    }

    let columns = get_all_columns(conn).await.unwrap_or_default();
    let functions = get_functions(conn).await.unwrap_or_default();
    let foreign_keys = get_foreign_keys(conn).await.unwrap_or_default();

    Ok(CompletionResponse {
        tables,
        columns,
        functions,
        foreign_keys,
    })
}
